### DISCUSSION TITLE
Setting quotas in FormR

### DISCUSSION THREAD

**Discussion by tbortolini** [rubenarslan/formr.org#519] [Q&A] [answered]
Hi everyone, I would like to know whether someone has solved the issue of setting quotas within a run (e.g., the maximum number of participants for a survey or a group).  

I've seen this previous suggestion, but we have run into the difficulties: https://groups.google.com/g/formr/c/TqmevSK_4lA/m/v5J1Gf4NAwAJ 

Is it still impossible to access the whole data frame from a survey? (https://groups.google.com/g/formr/c/f2BqbOTQ-PU/m/4QV_qN-BBQAJ).

Thank you very much!

**Comment by chei96** [comment, marked as answer]
I have used quotas in a formr study before. [Here is an example google spreadsheet](https://docs.google.com/spreadsheets/d/19U5g6pOAVi4Qi6I6mAclqwBFvj6nk6t_vM7lhwoc_3Y/edit?usp=sharing).

It works the way that:
1) people answer a secreening questionaire at the beginning of the run (e.g., age, gender, education)
2) You use your formr API to access your study data and then calculate the current number of participants in each quota group. I created separate variables for each quota like this:
`formr_connect() #use API to connect to formr`
`n_vps <- formr_raw_results("yourstudyname")`
`n_vps <- n_vps[complete.cases(n_vps[,c("session","sex","age","education","VARIABLES THAT INDICATE FULL PARTICIPATION")]),]`
`nrow(n_vps[n_vps$sex == 1,])`

3) use a block item with a show-if condition. Show that block item if the respective quota is full (i.e., if the number of participants is equal or lager the number of participants you defined for the quota). The showif could look like this:
`screen_sex == 3 & divers >= 1000|` 
`screen_sex == 1 & men >= 1000 |`
`screen_sex == 2 & women >= 1000 |`
`screen_age >= 18 & screen_age <20 & age_18 >= 1000 |`
`screen_age >= 20 & screen_age <29 & age_20 >= 1000 |`
`screen_age >= 30 & screen_age <39 & age_30 >= 1000 |`
`screen_age >= 40 & screen_age <49 & age_40 >= 1000 |`
`screen_age >= 50 & screen_age <59 & age_50 >= 1000 |`
`screen_age >= 60 & screen_age <69 & age_60 >= 1000 |`
`screen_education == 1 & edu1 >= 1000 |`
`screen_education == 2 & edu2 >= 1000 |`
`screen_education == 3 & edu3 >= 1000 |`
`screen_education == 4 & edu4 >= 1000 `

Replace the 1000 at the end of each line with the maximum number of participants in that quota group

This was one of my first formr studies, so there are some things to be considered:
 <ol>
  <li>I had the screening questions at the beginning and then, again, demographic questions in the survey. That was not optimal, as particiapnts had to answer that stuff twice. It would be better to just acess the screening answers in your survey data (e.g., use a hidden field in the study survey, in which you access the screening survey data). </li>
  <li>I loaded the current study data in the value field of each quota variable to calculate the current number of participants in that quota. This is not directly a problem but it might be possible to calculate the quotas by accessing the current study data only once, which would probably use less server resources.</li>
  <li>If many people try to access the survey at the same time (e.g., you use a panel provider where hundreds of users are invited at the same time), this might lead to an overfilling of the quota, as new participants can register while others, who will already fill the quota, are still participating. There might be ways to prevent that (e.g., place a block item directly at the beginning of the screening survey and use a show-if condition in which you access the number of participants who started study in the last x min. Show them text like "Too many people are currently trying to access the study, please come back later"</li>
</ol>  

I hope that helps anyone
Kindly 
Chris

**Reply by tbortolini** [reply]
Dear Chris, that was extremely helpful. Thank you very much for your collaboration. 

We'll try to implement a similar approach from our side.
